### ISSUE TITLE
Jellyfin support?

### ISSUE THREAD

**Issue by CrazyWolf13** [memen45/SubMusic#75]
Jellyfin support would be very cool as it basically replaces the paid plex.

**Comment by memen45** [comment]
Thanks for your feedback and suggestion! There have been more requests to implement Jellyfin support, so I thought it would be useful to start documenting the required information. If you have a server available to test and find out how to obtain the below information from the API, please feel free to share your insights.

- [ ] how to obtain access to the API? Authentication flow, username password, keys.
- [ ] how to obtain the required data? Below you can find the current features that are supported by SubMusic that require communication with the backend server. 

| provider method  	| Jellyfin API 	| description                                                  	| Content-Type     	|
|------------------	|----------	|--------------------------------------------------------------	|------------------	|
| ping             	| `/` 	| returns an object with server version                        	| application/json 	|
| recordPlay       	|          	| submit a play                                                	| application/json 	|
| getAllPlaylists  	|          	| returns array of all playlists available for the user        	| application/json 	|
| getPlaylist      	|          	| returns an array of one playlist object with {id}            	| application/json 	|
| getPlaylistSongs 	|          	| returns an array of songs on the playlist with id            	| application/json 	|
| getRefId         	|          	| returns a refId for a song by {id} (this downloads the song) 	| audio/*          	|
| getArtwork       	|          	| returns a BitmapResource for a song id                       	| image/*          	|
| getAllPodcasts   	|          	| returns array of all podcasts available for the user         	| application/json 	|
| getPodcast       	|          	| returns an array of one podcast object with {id}             	| application/json 	|
| getEpisodes      	|          	| returns array of episodes for podcast with {id}              	| application/json 	|

The Jellyfin API is documented [here](https://api.jellyfin.org/).

**Comment by mattgit** [comment]
> Thanks for your feedback and suggestion! There have been more requests to implement Jellyfin support, so I thought it would be useful to start documenting the required information. If you have a server available to test and find out how to obtain the below information from the API, please feel free to share your insights.
> 
> * [ ]  how to obtain access to the API? Authentication flow, username password, keys.
> * [ ]  how to obtain the required data? Below you can find the current features that are supported by SubMusic that require communication with the backend server.
> 
> provider method	Jellyfin API	description	Content-Type
> ping	`/`	returns an object with server version	application/json

https://api.jellyfin.org/#tag/System/operation/GetSystemInfo
Resp: Version

> recordPlay		submit a play	application/json

meaning?

> getAllPlaylists		returns array of all playlists available for the user	application/json

https://api.jellyfin.org/#tag/Items includeItemTypes "Playlist"

> getPlaylist		returns an array of one playlist object with {id}	application/json

What's taking in input?

> getPlaylistSongs		returns an array of songs on the playlist with id	application/json

https://api.jellyfin.org/#tag/Playlists/operation/GetPlaylistItems

> getRefId		returns a refId for a song by {id} (this downloads the song)	audio/*

https://api.jellyfin.org/#tag/Library/operation/GetDownload ?

> getArtwork		returns a BitmapResource for a song id	image/*

https://api.jellyfin.org/#tag/Image/operation/GetItemImage

> getAllPodcasts		returns array of all podcasts available for the user	application/json

https://api.jellyfin.org/#tag/Items includeItemTypes "Series" => some kind of? I'm not sure here

> getPodcast		returns an array of one podcast object with {id}	application/json

https://api.jellyfin.org/#tag/Items ids 

> getEpisodes		returns array of episodes for podcast with {id}	application/json

https://api.jellyfin.org/#tag/TvShows/operation/GetEpisodes
> The Jellyfin API is documented [here](https://api.jellyfin.org/).

trying

thank you, this would be great as Jellyfin has everything you need for free and it's really opensource.

**Comment by memen45** [comment]
Thanks for your help! I see that the Items endpoint is the one to use for discovering the playlists, couldn't find it before.

Found the authentication method [here](https://api.jellyfin.org/#tag/User/operation/AuthenticateUserByName): `/Users/AuthenticateByName`. Username and password are sent in plain text, response contains the AuthToken key. This token can be used in Authorization header as described [here](https://gist.github.com/nielsvanvelzen/ea047d9028f676185832e51ffaf12a6f).

The recordPlay function submits a timestamp + song id to record that the song has been played. This is typically used for statistics on the server side. Still unclear which endpoint could achieve it.

I've been searching for podcast support. Unfortunately, it looks like Jellyfin does not support it yet, see more details [here](https://features.jellyfin.org/posts/48/podcast-support). For SubMusic, this is not a problem at all, as the Plex backend also lacks podcast support. For normal playlists you can still enable podcast mode if needed.

**Comment by warreth** [comment]
I am willing to work on this, but @memen45 are you still maintaining this project?

**Comment by memen45** [comment]
Last year I already built Emby and Jellyfin support in SubMusic. There are two reasons this isn't released yet.

### Emby Connect 
This feature is not tested yet and poorly documented. In terms of good UX I would prefer the Emby Connect feature over the Nextcloud style of connecting (fill in server url, Username and password). However, that would require either creating a separate app, like SubMusic for Plex, or adding a Connect feature into the SubMusic app. The latter could create confusion about whether app settings are actually used to connect to the server, or it is happening through the Emby Connect.
### Emby Podcast
This feature is not tested yet. Lacking documentation in Emby, so have to test against a (paid) Emby server with the podcast feature, for which I did not have time yet.
### Integration of monetization
Probably not the most popular topic, but maintaining these apps costs a lot of time, which unfortunately is not free for me. In order to be able to dedicate time to the development of SubMusic, I am thinking about ways to monetize it, similar to SubMusic for Plex. However, I also want to contribute to the open source projects like Nextcloud, Ampache and Airsonic by enabling their use on these 'exotic' Garmin devices. I have not decided on a right balance between the two. One way could be to have certain premium/advanced features behind subscription, like auto-sync, bitrate choice, etc. Or limit the number of playlists/songs one can sync with the free tier. I do not know whether there is a good solution for this, so that is why SubMusic itself is still 'as-is'. Suggestions are welcome ofcourse.

**Comment by teranex** [comment]
Personally I would be fine with a free version which allows for 50 songs (as an example) and then a paid version if you want to download more. Or a version that can be tested for 2 weeks for example. I'm a big fan of open source, but I'm also fully aware how much time it takes to develop something, so I have no problem with paying for nice apps.

**Comment by warreth** [comment]
i would just release the garmin app for a one-time fee, like 5 bucks.

I, and others would gladly pay for it!
a subscription would not sell well, because its againts the principle of jellyfin.

**Comment by warreth** [comment]
@memen45 Have you decided whether you’ll release it? If not, I’ll fork it and implement it myself - I’d really like to be able to use Jellyfin.

**Comment by memen45** [comment]
Yes, it will definitely be released. It takes a bit of time to finish the update completely. In the mean time, I could provide you a dev version for you to sideload. Would be great if you could help with some feedback. Please reach out to me through the 'Contact Developer' option on Garmin Connect IQ > SubMusic app page.

You can of course fork the SubMusic repo, however, the code is not representing the current released state. This is due to a large overhaul of i.a. memory structure and management that was built in the mean time, as well as many bug fixes and stability improvements. Therefore, I would not recommend building upon the old version that is shown here.
